### PR TITLE
LDAP protocol parser v3

### PR DESCRIPTION
**Pull request by glongo**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)
- [x] I have created a ticket at
      https://redmine.openinfosecfoundation.org/projects/suricata/issues
      (if applicable)

Link to ticket: https://redmine.openinfosecfoundation.org/issues/1199

Describe changes:
- Rebase of #11087 

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1860

### CODE REVIEW

**Comment by catenacyber** [comment]
Thanks for this work Giuseppe.

Taking a step back, my big question is :
Are there detection keywords ? frames ? (frames may be cheap to add and powerful)

Could you add the redmine ticket number in a commit message ?
(@jasonish is not CI supposed to be red because of this missing ? )

Looks like there is some documentation to add...

**Review by catenacyber — CHANGES_REQUESTED** [review]
I did not re-look at the code, but this needs some more polishing :-)

**Comment by jasonish** [comment]
> Thanks for this work Giuseppe.
> 
> 
> 
> Taking a step back, my big question is :
> 
> Are there detection keywords ? frames ? (frames may be cheap to add and powerful)
> 
> 
> 
> Could you add the redmine ticket number in a commit message ?
> 
> (@jasonish is not CI supposed to be red because of this missing ? )

There's no check for that.

**Comment by jasonish** [comment]
When looking at a pcap with:
- bind
- search
- unbind

Suricata doesn't log in that order. Can we fix that?

File: rust/src/ldap/logger.rs [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+        js.set_string_from_bytes("value", &value[..])?;
+    }
+    js.close()?;
+    Ok(())
+}
+
+fn log_ldap_result(msg: &LdapResult, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.set_string("result_code", &msg.result_code.to_string())?;
+    js.set_string("matched_dn", &msg.matched_dn.0)?;
+    js.set_string("message", &msg.diagnostic_message.0)?;
```
Comment (jasonish) [inline rust/src/ldap/logger.rs:324]: These 2 seem empty a lot. I wonder if we should only log them when there is data.

**Comment by jasonish** [comment]
I find the logging very verbose, which on one hand is nice, but I'm going to guess if this is deployed on a network with LDAP, it might just lead to overload. 

Do we have any idea what kind of load this would generate on a network where LDAP is used?

Can we reduce the datapoints logged?

Taking a quick look at a PCAP I pulled from cloudshark:

```
  "ldap": {
    "request": {
      "message_id": 5,
      "operation": "search_request",
      "search_request": {
        "base_object": "cn=Users,dc=cloudshark-a,dc=example,dc=com",
        "scope": 2,
        "deref_alias": 0,
        "size_limit": 0,
        "time_limit": 0,
        "types_only": false,
        "attributes": [
          "objectClass",
          "name",
          "gidNumber",
          "objectGUID",
          "objectSID",
          "whenChanged",
          "uSNChanged",
          "groupType"
        ]
      },
      "controls": [
        {
          "control_type": "1.2.840.113556.1.4.319",
          "criticality": false,
          "control_value": "3006020203e80400"
        }
      ]
    },
    "responses": [
      {
        "operation": "search_result_done",
        "search_result_done": {
          "result_code": "success",
          "matched_dn": "",
          "message": ""
        },
        "controls": [
          {
            "control_type": "1.2.840.113556.1.4.319",
            "criticality": false,
            "control_value": "3084000000050201000400"
          }
        ]
      }
    ]
  }
```

I think we have to reason about:
- 1:1 parity with keywords. Thats a lot right there, and that's just one message type
- What could be eliminated? Is our goal to provide as much context as Wireshark? Or should going to PCAP still be the way to get full fidelity.

What I'm getting at is this could be a lot of data, perhaps too much. Do rule writers need access to every buffer?  Or would frames be a better approach for rule writers, then the log being a much more succint summary of what happened.

**Comment by glongo** [comment]
> When looking at a pcap with:
> 
>     * bind
> 
>     * search
> 
>     * unbind
> 
> 
> Suricata doesn't log in that order. Can we fix that?

How they are ordered?

**Comment by jasonish** [comment]
> > When looking at a pcap with:
> > ```
> > * bind
> > 
> > * search
> > 
> > * unbind
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Suricata doesn't log in that order. Can we fix that?
> 
> How they are ordered?

In the pcap we have:
- bind request
- bind response
- search
- unbind request

In the eve output we have:
- search
- unbind
- bind
[ldap-and-search.pcap.gz](https://github.com/user-attachments/files/15795276/ldap-and-search.pcap.gz)

**Comment by glongo** [comment]
Replaced with https://github.com/OISF/suricata/pull/11349

**Pull request closed without merging**